geowebcache/mbtiles/src/main/java/org/geowebcache/mbtiles/layer/MBTilesLayer.java

```diff
@@ -0,0 +1,349 @@
+
+        if (tileSize <= 0) {
+            tileSize = DEFAULT_TILE_SIZE;
+        }
+
+        switch (tileSize) {
+            case TILE_SIZE_256:
+                gridSet = defaultGridSet.worldEpsg3857();
+                break;
+            case TILE_SIZE_512:
+                gridSet = defaultGridSet.worldEpsg3857x2();
+                break;
+            default:
+                throw new IllegalArgumentException("Unsupported tileSize: " + tileSize);
+        }
+        Integer minZoom = cacheInfo.getMinZoom();
+        Integer maxZoom = cacheInfo.getMaxZoom();
+
+        GridSubset subSet =
+                GridSubsetFactory.createGridSubSet(gridSet, this.layerBounds, minZoom, maxZoom);
+
+        HashMap<String, GridSubset> subsets = new HashMap<>();
+        subsets.put(gridSet.getName(), subSet);
+        return subsets;
+    }
+
+    /** @see org.geowebcache.layer.TileLayer#getTile(org.geowebcache.conveyor.ConveyorTile) */
+    @Override
+    public ConveyorTile getTile(final ConveyorTile tile)
+            throws IOException, OutsideCoverageException {
+
+        long[] tileIndex = tile.getTileIndex();
+        int zl = (int) tileIndex[2];
+        int row = (int) tileIndex[1];
+        int column = (int) tileIndex[0];
+        MBTilesTile loadedTile = mbTilesFile.loadTile(zl, column, row);
+        byte[] content = loadedTile.getData();
+        if (content != null) {
+            if (cacheInfo.getFormat() == MBTilesMetadata.t_format.PBF) {
+                content = getPbfFromTile(content);
+            }
+
+            tile.setBlob(new ByteArrayResource(content));
+            tile.setCacheResult(CacheResult.HIT);
+        } else {
+            tile.setCacheResult(CacheResult.MISS);
+            throw new OutsideCoverageException(tile.getTileIndex(), 0, 0);
+        }
+
+        saveExpirationInformation((int) (tile.getExpiresHeader() / 1000));
+
+        return tile;
+    }
+
+    private byte[] getPbfFromTile(byte[] raw) throws IOException {
+        if (!UNZIP_CONTENT) {
+            return raw;
+        }
+        byte[] byteArray = null;
+        try (ByteArrayOutputStream out = new ByteArrayOutputStream();
+                ByteArrayInputStream bin = new ByteArrayInputStream(raw);
+                InflaterInputStream in = new GZIPInputStream(bin)) {
+            byte[] buffer = new byte[1024];
+            int noRead;
+            while ((noRead = in.read(buffer)) != -1) {
+                out.write(buffer, 0, noRead);
+            }
+            byteArray = out.toByteArray();
+        }
+        return byteArray;
+    }
+
+    protected void saveExpirationInformation(int backendExpire) {
+        this.saveExpirationHeaders = false;
+
+        try {
+            if (getExpireCache(0) == GWCVars.CACHE_USE_WMS_BACKEND_VALUE) {
+                if (backendExpire == -1) {
+                    this.expireCacheList.set(0, new ExpirationRule(0, 7200));
+                    log.error(
```

Comment:
Logging at error level seems excessive here, this will be logged for every tile request. Guard and log at a lower level IMHO. Same for the log below.

Reply:
Also guard the trace logs, as this is a method called for each tile.

Reply:
Sounds good. Will fix it (I took that piece of code from ArcGISCacheLayer :-) )